core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -450,24 +457,54 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
                         logger.debug("connected to node [{}]", node);
                     }
                     transportServiceAdapter.onNodeConnected(node);
+                    success = true;
                 } catch (ConnectTransportException e) {
                     throw e;
                 } catch (Exception e) {
                     throw new ConnectTransportException(node, "general node connection failure", e);
+                } finally {
+                    if (success == false) {
+                        connectedNodes.remove(node);
```

Comment:
yeah, that's a good one. I made it such so that exceptions on the `transportServiceAdapter.onNodeConnected(node);` will mean an unsuccessful connection, as will any exception that bubbles out of this method.